### PR TITLE
admin endpoint for listing and revoking tokens

### LINKED ISSUE #1000: Admin endpoints to handle tokens
let's add admin endpoints for:
- listing tokens, optionally filtering by token, hash, outpoint, or txid. Returns the list of tokens with their outpoints and expiry date
- revoking tokens, optionally filtering by token, hash, outpoint or txid. Forces the cancelation of the token (hash) from the token cache

### PR DESCRIPTION
**Pull request by bitcoin-coder-bob**
closes https://github.com/arkade-os/arkd/issues/1000

  - Add ListTokens (GET /v1/admin/tokens) and RevokeTokens (POST /v1/admin/tokens/revoke) to AdminService                                                                                                  
  - Both accept optional filters: token (base64 auth token), hash (hex), outpoint (txid:vout), txid — filters are ANDed when combined
  - ListTokens returns matching entries sorted oldest-first, capped at 50 to bound response size so  a sort of DoS prevention in that we dont return a giant amount potentially                                                                                                       
  - RevokeTokens requires at least one filter to prevent accidental full-cache wipe                                                                                                                        
  - Token filter verifies schnorr signature (skips expiry so admins can inspect/revoke expired tokens)                                                                                                     
  - Outpoint filter is validated and normalized before use                                                                                                                                                 
  - Both endpoints gated behind manager entity permissions (read/write)

## Summary by CodeRabbit

* **New Features**
  * Added admin endpoints to list and revoke tokens (POST /v1/admin/tokens and POST /v1/admin/tokens/revoke). Listing returns token entries including outpoints and expiry.

* **Bug Fixes**
  * Improved validation and normalization for token/hash/outpoint/txid inputs; revoke now rejects requests with no filters.

* **Tests**
  * Added comprehensive unit tests for listing, revocation, filtering, normalization, validation, and edge cases.

### CODE REVIEW

**Comment by bitcoin-coder-bob** [comment]
@coderabbitai review

File: internal/interface/grpc/handlers/adminservice.go [hunk 1]
```diff
@@ -502,6 +505,46 @@ func (a *adminHandler) RevokeAuth(
...
 }
 
+func (a *adminHandler) ListTokens(
+	ctx context.Context, req *arkv1.ListTokensRequest,
+) (*arkv1.ListTokensResponse, error) {
+	tokens, err := a.indexerService.ListTokens(
+		ctx, req.GetToken(), req.GetHash(), req.GetOutpoint(), req.GetTxid(),
+	)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "%s", err.Error())
```
[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply internal/interface/grpc/handlers/adminservice.go]: done in: [ad30c99](https://github.com/arkade-os/arkd/pull/1001/commits/ad30c9999f2225a1b3908a4c9acd50e64183148f)

[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply internal/interface/grpc/handlers/adminservice.go]: done in: [ad30c99](https://github.com/arkade-os/arkd/pull/1001/commits/ad30c9999f2225a1b3908a4c9acd50e64183148f)

File: internal/core/application/indexer.go [hunk 2]
```diff
@@ -949,6 +951,86 @@ func (i *indexerService) validateAuthToken(authToken string) (string, error) {
...
+	h, err := i.resolveTokenFilter(token, hash)
+	if err != nil {
+		return 0, err
+	}
+	op, err := normalizeOutpoint(outpoint)
+	if err != nil {
+		return 0, err
+	}
+	return i.tokenCache.revoke(h, op, txid), nil
+}
```
[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer.go:1029]: done in: [ad30c99](https://github.com/arkade-os/arkd/pull/1001/commits/ad30c9999f2225a1b3908a4c9acd50e64183148f)

File: internal/core/application/indexer.go [hunk 3]
```diff
@@ -949,6 +951,86 @@ func (i *indexerService) validateAuthToken(authToken string) (string, error) {
...
+// Returns empty string if input is empty.
+func normalizeOutpoint(outpoint string) (string, error) {
+	if outpoint == "" {
+		return "", nil
+	}
+	var op Outpoint
+	if err := op.FromString(outpoint); err != nil {
+		return "", fmt.Errorf("invalid outpoint filter: %w", err)
+	}
+	return op.String(), nil
```
[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer.go:997]: done in: [ad30c99](https://github.com/arkade-os/arkd/pull/1001/commits/ad30c9999f2225a1b3908a4c9acd50e64183148f)

File: api-spec/protobuf/ark/v1/admin.proto [hunk 4]
```diff
@@ -110,6 +110,17 @@ service AdminService {
       body: "*"
     };
   }
+  rpc ListTokens(ListTokensRequest) returns (ListTokensResponse) {
+    option (meshapi.gateway.http) = {
+      get: "/v1/admin/tokens"
+    };
```
[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply api-spec/protobuf/ark/v1/admin.proto:117]: done in: [ad30c99](https://github.com/arkade-os/arkd/pull/1001/commits/ad30c9999f2225a1b3908a4c9acd50e64183148f)

File: api-spec/openapi/swagger/ark/v1/admin.openapi.json [hunk 5]
```diff
@@ -1512,6 +1648,37 @@
...
+          },
+          "token": {
+            "type": "string",
+            "description": "base64 auth token"
+          },
+          "txid": {
+            "type": "string"
+          }
+        }
+      },
```
[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply api-spec/openapi/swagger/ark/v1/admin.openapi.json:1648]: This file is auto-generated by ` make proto` so any handmade edits get overwritten on the next generation. The constraint is already enforced at two layers. The gRPC handler and the service layer both reject empty filters.

File: internal/interface/grpc/handlers/adminservice.go [hunk 6]
```diff
@@ -26,16 +27,19 @@ import (
 
 type adminHandler struct {
 	adminService    application.AdminService
+	indexerService  application.IndexerService
```
Comment (altafan) [inline internal/interface/grpc/handlers/adminservice.go]: let's define here a private interface just to reduce the visibility to only the methods we need:
```
type tokenAdminInterface interface {
    ListTokens(ctx context.Context, token, hash, outpoint, txid string) ([]application.TokenEntry, error)
	RevokeTokens(ctx context.Context, token, hash, outpoint, txid string) (int, error)
}
```


```suggestion
	tokenAdminService  tokenAdminInterface
```

File: internal/interface/grpc/handlers/adminservice.go [hunk 7]
```diff
@@ -502,6 +506,52 @@ func (a *adminHandler) RevokeAuth(
 	}, nil
 }
 
+func (a *adminHandler) ListTokens(
```
Comment (altafan) [inline internal/interface/grpc/handlers/adminservice.go:524]: Please add a TODO above these 2 methods to not forget to be moved to the indexer's own admin interface when we detach it. Having them here is ok for now, but it's just a workaround

File: internal/interface/grpc/handlers/adminservice_test.go [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+	arkv1 "github.com/arkade-os/arkd/api-spec/protobuf/gen/ark/v1"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func TestRevokeTokensRequiresFilter(t *testing.T) {
+	handler := &adminHandler{}
+
+	_, err := handler.RevokeTokens(context.Background(), &arkv1.RevokeTokensRequest{})
```
Comment (altafan) [inline internal/interface/grpc/handlers/adminservice_test.go]:
```suggestion
	_, err := handler.RevokeTokens(t.Context(), &arkv1.RevokeTokensRequest{})
```

File: internal/core/application/indexer.go [hunk 9]
```diff
@@ -949,6 +954,92 @@ func (i *indexerService) validateAuthToken(authToken string) (string, error) {
...
+	sigBytes := tokenBytes[40:]
+
+	msgHash := chainhash.HashB(msg)
+	sig, err := schnorr.ParseSignature(sigBytes)
+	if err != nil {
+		return "", fmt.Errorf("%w: failed to parse auth token signature: %w", ErrInvalidInput, err)
+	}
+	if !sig.Verify(msgHash, i.authPrvkey.PubKey()) {
+		return "", fmt.Errorf("%w: signature verification failed", ErrInvalidInput)
+	}
```
Comment (altafan) [inline internal/core/application/indexer.go]: no need to do verification here, let's drop it. We assume we call ListTokens/RevokeTokens for valid ones. If the hash is not found in cache we eventually get error, but i don't see any real reason to check the sig here

File: internal/core/application/indexer.go [hunk 10]
```diff
@@ -949,6 +954,92 @@ func (i *indexerService) validateAuthToken(authToken string) (string, error) {
...
+	_ context.Context, token, hash, outpoint, txid string,
+) ([]TokenEntry, error) {
+	h, err := i.resolveTokenFilter(token, hash)
+	if err != nil {
+		return nil, err
+	}
+	op, err := normalizeOutpoint(outpoint)
+	if err != nil {
+		return nil, err
+	}
```
Comment (altafan) [inline internal/core/application/indexer.go:1010]: if no filter is provided we should return an error instead this would return an empty list of tokens?

File: internal/core/application/indexer_admin_test.go [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+		authTokenTTL: defaultAuthTokenTTL,
+		tokenCache:   newTokenCache(defaultAuthTokenTTL),
+	}
+	t.Cleanup(svc.tokenCache.close)
+
+	outpoints := []Outpoint{
+		{Txid: "aabbcc1100000000000000000000000000000000000000000000000000000000", VOut: 0},
+	}
+
+	t.Run("extracts hash from valid token", func(t *testing.T) {
```
Comment (altafan) [inline internal/core/application/indexer_admin_test.go]: this is enough as name
```suggestion
	t.Run("valid token", func(t *testing.T) {
```

File: internal/core/application/indexer_admin_test.go [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+		expectedHash, err := svc.validateAuthToken(token)
+		require.NoError(t, err)
+
+		// extractTokenHash should return the same hash.
+		hash, err := svc.extractTokenHash(token)
+		require.NoError(t, err)
+		require.Equal(t, expectedHash, hash)
+	})
+
+	t.Run("extracts hash from expired token", func(t *testing.T) {
```
Comment (altafan) [inline internal/core/application/indexer_admin_test.go]:
```suggestion
	t.Run("expired token", func(t *testing.T) {
```

File: internal/core/application/indexer_admin_test.go [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+		// validateAuthToken should fail (expired).
+		_, err = shortSvc.validateAuthToken(token)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "expired")
+
+		// extractTokenHash should still succeed.
+		hash, err := shortSvc.extractTokenHash(token)
+		require.NoError(t, err)
+		require.NotEmpty(t, hash)
+	})
```
Comment (altafan) [inline internal/core/application/indexer_admin_test.go]: let's move these tests into `t.Run("valid", func(t *testing.T))`

File: internal/core/application/indexer_admin_test.go [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+		}
+		t.Cleanup(publicSvc.tokenCache.close)
+
+		token, err := svc.createAuthToken(outpoints)
+		require.NoError(t, err)
+
+		_, err = publicSvc.extractTokenHash(token)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "public exposure")
+	})
```
Comment (altafan) [inline internal/core/application/indexer_admin_test.go]: let's move these tests into `t.Run("invalid", func(t *testing.T))`

File: internal/core/application/indexer_admin_test.go [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+		require.NotEmpty(t, h)
+	})
+
+	t.Run("propagates token error", func(t *testing.T) {
+		_, err := svc.resolveTokenFilter("bad-token!!!", "")
+		require.Error(t, err)
+	})
+}
+
+func TestNormalizeOutpoint(t *testing.T) {
```
Comment (altafan) [inline internal/core/application/indexer_admin_test.go:144]: also here we can split test cases into valid and invalid

File: internal/core/application/indexer_exposure_test.go [hunk 16]
```diff
@@ -920,6 +920,181 @@ func TestStripSignerSignatures(t *testing.T) {
 	})
 }
 
+func TestListAndRevokeTokens(t *testing.T) {
```
Comment (altafan) [inline internal/core/application/indexer_exposure_test.go]: let's split this test into TestListTokens and TestRevokeTokens.

Also here, we can also split the test cases by `valid` and `invalid` for each test

File: internal/core/application/token_cache.go [hunk 17]
```diff
@@ -119,3 +128,143 @@ func (c *tokenCache) getTxids(hash string) (map[string]struct{}, bool) {
...
+// Results are sorted oldest-first and capped at maxTokenListSize to
+// bound response size.
+func (c *tokenCache) list(hash, outpointStr, txid string) []TokenEntry {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	now := time.Now()
+	var result []TokenEntry
+
+	for h, outpoints := range c.outpointsByHash {
```
Comment (altafan) [inline internal/core/application/token_cache.go:142]: looping over outpointsByHash unconditionally can be extremely inefficient in case the hash filter is provided, while it should just be O(1). That should be done only in case outpoint or txid filter is provided.

```suggestion
func (c *tokenCache) list(hash, outpointStr, txid string) []TokenEntry {
	if hash != "" {
       indexedOutpoints, expiry, ok := c.getOutpoints(hash)
       if !ok {
           return nil, nil // Or not found error?
       }
       
       outpoints := make([]string, 0, len(indexedOutpoints)
       for outpoint := range indexedOutpoints {
           outpoints = append(outpoints, outpoint)
       }
       return []TokenEntry{{
          Hash: hash,
          Outpoints: outpoints,
          ExpiresAt: expiry,
       }}
    }
    
    c.mu.RLock()
	defer c.mu.RUnlock()

	now := time.Now()
	var result []TokenEntry

	for h, outpoints := range c.outpointsByHash {
```

we could remove checks over hash in this loop

File: internal/core/application/token_cache.go [hunk 18]
```diff
@@ -119,3 +128,143 @@ func (c *tokenCache) getTxids(hash string) (map[string]struct{}, bool) {
...
+
+		// Collect outpoints as sorted strings for deterministic output.
+		entry := TokenEntry{
+			Hash:      h,
+			ExpiresAt: expiresAt,
+		}
+		for op := range outpoints {
+			entry.Outpoints = append(entry.Outpoints, op.String())
+		}
+		sort.Strings(entry.Outpoints)
```
Comment (altafan) [inline internal/core/application/token_cache.go]: no need to sort, we can drop

File: internal/core/application/token_cache.go [hunk 19]
```diff
@@ -119,3 +128,143 @@ func (c *tokenCache) getTxids(hash string) (map[string]struct{}, bool) {
...
+	}
+
+	// Sort oldest first (earliest expiry = earliest creation).
+	sort.Slice(result, func(i, j int) bool {
+		return result[i].ExpiresAt.Before(result[j].ExpiresAt)
+	})
+
+	if len(result) > maxTokenListSize {
+		result = result[:maxTokenListSize]
+	}
```
Comment (altafan) [inline internal/core/application/token_cache.go]: these checks seem overkill, let's drop them

File: internal/core/application/token_cache.go [hunk 20]
```diff
@@ -119,3 +128,143 @@ func (c *tokenCache) getTxids(hash string) (map[string]struct{}, bool) {
...
+
+	return result
+}
+
+// revoke deletes non-expired token entries matching the given filters
+// and returns the number of entries removed. Filters are ANDed.
+func (c *tokenCache) revoke(hash, outpointStr, txid string) int {
```
Comment (altafan) [inline internal/core/application/token_cache.go:224]: same as for list(), let's skip looping if the hash filter is provided and leverage getTxids in this case (we don't need expiry)

File: internal/core/application/token_cache_test.go [hunk 21]
```diff
@@ -135,4 +136,249 @@ func TestTokenCache(t *testing.T) {
...
+				setup: func(c *tokenCache) {
+					c.add("hash1", []Outpoint{op1}, time.Now())
+					c.add("hash2", []Outpoint{op3}, time.Now()) // op3 shares txid with op1
+				},
+				assert: func(t *testing.T, c *tokenCache) {
+					entries := c.list("hash1", "", op1.Txid)
+					require.Len(t, entries, 1)
+					require.Equal(t, "hash1", entries[0].Hash)
+				},
+			},
```
Comment (altafan) [inline internal/core/application/token_cache_test.go]: this is not an interesting test case. We won't use combined filters. Ideally they are all mutually exclusive and we have many of them to facilitate the lookup for tokens

File: internal/core/application/token_cache_test.go [hunk 22]
```diff
@@ -135,4 +136,249 @@ func TestTokenCache(t *testing.T) {
...
+							VOut: 0,
+						}
+						c.add(fmt.Sprintf("hash%d", i), []Outpoint{op}, now)
+					}
+				},
+				assert: func(t *testing.T, c *tokenCache) {
+					entries := c.list("", "", "")
+					require.Len(t, entries, maxTokenListSize)
+				},
+			},
```
Comment (altafan) [inline internal/core/application/token_cache_test.go]: after previous comments, these could be removed

File: internal/core/application/token_cache_test.go [hunk 23]
```diff
@@ -135,4 +136,249 @@ func TestTokenCache(t *testing.T) {
...
+				setup: func(c *tokenCache) {
+					c.add("hash1", []Outpoint{op1}, time.Now())
+					c.add("hash2", []Outpoint{op2}, time.Now())
+				},
+				assert: func(t *testing.T, c *tokenCache) {
+					entries := c.list("", "", op1.Txid)
+					require.Len(t, entries, 1)
+					require.Equal(t, "hash1", entries[0].Hash)
+				},
+			},
```
Comment (altafan) [inline internal/core/application/token_cache_test.go]: in these tests, it would be interesting to see that if 2 different tokens include the same outpoint/txid, they are both returned:
```suggestion
				name: "filter by outpoint",
				setup: func(c *tokenCache) {
					c.add("hash1", []Outpoint{op1, op2}, time.Now())
					c.add("hash2", []Outpoint{op2, op3}, time.Now())
					c.add("hash3", []Outpoint{op1, op3}, time.Now())
				},
				assert: func(t *testing.T, c *tokenCache) {
					entries := c.list("", op2.String(), "")
					require.Len(t, entries, 2)
					// Or find a better way to check both ahshes are in the list
					require.Equal(t, "hash1", entries[0].Hash)
					require.Equal(t, "hash2", entries[1].Hash)
				},
			},
			{
				name: "filter by txid",
				setup: func(c *tokenCache) {
					c.add("hash1", []Outpoint{op1}, time.Now())
					c.add("hash2", []Outpoint{op1, op2}, time.Now())
					c.add("hash2", []Outpoint{op2}, time.Now())
				},
				assert: func(t *testing.T, c *tokenCache) {
					entries := c.list("", "", op1.Txid)
					require.Len(t, entries, 2)
					require.Equal(t, "hash1", entries[0].Hash)
					require.Equal(t, "hash2", entries[1].Hash)
				},
			},
```

File: internal/core/application/token_cache_test.go [hunk 24]
```diff
@@ -135,4 +136,249 @@ func TestTokenCache(t *testing.T) {
...
+					// hash1 gone
+					_, _, ok := c.getOutpoints("hash1")
+					require.False(t, ok)
+					// hash2 still there
+					_, _, ok = c.getOutpoints("hash2")
+					require.True(t, ok)
+				},
+			},
+			{
+				name: "by outpoint",
```
Comment (altafan) [inline internal/core/application/token_cache_test.go]: the considerations made for list apply also to this test. The test case below covers in total the case for the txid filer instead

File: internal/core/domain/vtxo.go [hunk 25]
```diff
@@ -23,11 +23,19 @@ func (k *Outpoint) FromString(s string) error {
...
 	}
-	k.Txid = parts[0]
+	txid := parts[0]
+	if len(txid) != 64 {
+		return fmt.Errorf("invalid txid length: expected 64 hex chars, got %d", len(txid))
+	}
+	txidBytes, err := hex.DecodeString(txid)
+	if err != nil {
+		return fmt.Errorf("invalid txid hex: %s", txid)
+	}
```
Comment (altafan) [inline internal/core/domain/vtxo.go]: let's switch the checks so we can return an error referring to 32 bytes instead of 64 hex chars:
```suggestion
	txidBytes, err := hex.DecodeString(txid)
	if err != nil {
		return fmt.Errorf("invalid txid hex: %s", txid)
	}
	if len(txidBytes) != 32 {
		return fmt.Errorf("invalid txid length: expected 32 bytes, got %d", len(txidBytes))
	}
```

File: internal/core/application/indexer.go (hunk 10 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer.go:1010]: I think no filter yields all tokens. which seems like a valid admin operation

File: internal/core/domain/vtxo.go (hunk 25 above)
Comment (bitcoin-coder-bob) [reply internal/core/domain/vtxo.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/token_cache_test.go (hunk 24 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/token_cache_test.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/token_cache_test.go (hunk 23 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/token_cache_test.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/token_cache_test.go (hunk 22 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/token_cache_test.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/token_cache_test.go (hunk 21 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/token_cache_test.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

Comment (bitcoin-coder-bob) [reply internal/core/application/token_cache_test.go]: removed it simply

File: internal/core/application/token_cache.go (hunk 20 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/token_cache.go:224]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d) please confirm this was done correctly

File: internal/core/application/token_cache.go (hunk 19 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/token_cache.go]: the idea was to enforce a max amount of tokens. possibly preventing some huge response that could be a DoS. Unlikely but that was the spirit. I removed them here now: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/token_cache.go (hunk 18 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/token_cache.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/token_cache.go (hunk 17 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/token_cache.go:142]: done by adding `listByHash` usage in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/indexer_exposure_test.go (hunk 16 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer_exposure_test.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/indexer_admin_test.go (hunk 15 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer_admin_test.go:144]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/indexer_admin_test.go (hunk 13 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer_admin_test.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/indexer_admin_test.go (hunk 14 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer_admin_test.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/indexer_admin_test.go (hunk 12 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer_admin_test.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/indexer_admin_test.go (hunk 11 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer_admin_test.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/indexer.go (hunk 9 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/interface/grpc/handlers/adminservice_test.go (hunk 8 above)
Comment (bitcoin-coder-bob) [reply internal/interface/grpc/handlers/adminservice_test.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/interface/grpc/handlers/adminservice.go (hunk 7 above)
Comment (bitcoin-coder-bob) [reply internal/interface/grpc/handlers/adminservice.go:524]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/interface/grpc/handlers/adminservice.go (hunk 6 above)
Comment (bitcoin-coder-bob) [reply internal/interface/grpc/handlers/adminservice.go]: done in: [164c679](https://github.com/arkade-os/arkd/pull/1001/commits/164c679c19e18991bcb19fb319f24401915fe94d)

File: internal/core/application/indexer.go [hunk 26]
```diff
@@ -949,6 +954,80 @@ func (i *indexerService) validateAuthToken(authToken string) (string, error) {
...
+	return i.tokenCache.list(h, op, txid), nil
```
[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer.go:1011]: @altafan had us remove the cap

File: internal/core/application/token_cache.go [hunk 27]
```diff
@@ -119,3 +125,174 @@ func (c *tokenCache) getTxids(hash string) (map[string]struct{}, bool) {
...
+		return c.listByHash(hash)
+	}
+
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	now := time.Now()
+	var result []TokenEntry
+
+	for h, outpoints := range c.outpointsByHash {
```
[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply internal/core/application/token_cache.go:142]: https://github.com/arkade-os/arkd/pull/1001#discussion_r3052820217   that is the filters are mutaully exclusive, we removed the combined filter

File: internal/core/application/indexer.go (hunk 10 above)
Comment (altafan) [reply internal/core/application/indexer.go:1010]: do we have a test that ensures this?

File: internal/core/application/token_cache.go (hunk 20 above)
Comment (altafan) [reply internal/core/application/token_cache.go:224]: yes, all good

File: internal/core/application/indexer.go (hunk 10 above)
Comment (bitcoin-coder-bob) [reply internal/core/application/indexer.go:1010]: yes. https://github.com/arkade-os/arkd/pull/1001/changes#diff-e2803c5c4c56c4d1a9d8fec1b9bc1ea7b28e177a53d6d7ce3f154dfe6d5468acR146-R154

**Review by altafan — APPROVED** [review]
